docs/export/export.md

```diff
@@ -631,10 +638,135 @@ We list here a history of the calling convention version numbers:
+We can start a backward-compatibility deprecation clock when we make a jaxlib release,
+even if we don’t force it to be the minimum allowed version.
+
+Let’s say that we need to add, delete, or change the semantics of a
+custom call target `T` used by the JAX lowering rules.
+Here is a possible chronology (for changing custom call targets
+that live in jaxlib):
+
+  1. Day “D - 1”, before the change. Say that the active internal JAX version is `0.4.31`
+     (the version of the next JAX and jaxlib releases).
+     The JAX lowering rules use a custom call `T`.
+  2. Day “D”, we add the new custom call target `T_NEW`.
+    We should create a new custom call target, and clean up the old
+    target roughly after 6 months, rather than updating `T` in place:
+       * See the example [PR #20997](https://github.com/google/jax/pull/20997)
+         implementing the steps below.
+       * We add the custom call target `T_NEW`.
+       * We change the JAX lowering rules that were previous using `T`,
+         to use `T_NEW`, conditionally as follows:
+
+        ```python
+        from jax._src import config
+        from jax._src.lib import version as jaxlib_version
+
+        def my_lowering_rule(ctx: LoweringRuleContext, ...):
+          lowering_parameters = ctx.module_context.lowering_parameters
+          forward_compat_mode = (lowering_parameters.for_export and
+                                 not lowering_parameters.export_ignore_forward_compatibility)
+          if forward_compat_mode or jaxlib_version < (0, 4, 31):
+            # this is the old lowering, using target T, while we
+            # are in forward compatibility mode for T, or we
+            # are in OSS and are using an old jaxlib.
+            return hlo.custom_call("T", ...)
+          else:
+            # This is the new lowering, using target T_NEW, for
+            # when we use a jaxlib with version `>= (0, 4, 31)`
+            # (or when this is internal usage), and also we are
+            # in JIT mode.
+            return hlo.custom_call("T_NEW", ...)
+        ```
+       * Note that the forward compatibility mode is always false in JIT mode
+         or if the user passes `--jax_export_ignore_forward_compatibility=true`
+       * We add `T_NEW` to the list of
+         [`_CUSTOM_CALL_TARGETS_GUARANTEED_STABLE`](https://github.com/search?q=repo%3Agoogle%2Fjax++%22_CUSTOM_CALL_TARGETS_GUARANTEED_STABLE+%3D%22+path%3A_export.py&amp%3Btype=code&type=code)
+         in `_export.py`.
+  3. Day “D + 21” (end of forward compatibility window; can be even later than 21 days):
+    We remove the `forward_compat_mode` in the lowering code, so now exporting
+    will start using the new custom call target `T_NEW` as long as we are using a new `jaxlib`.
+       * We add a backwards compatibility test for `T_NEW`.
+  4. Day "RELEASE > D" (the first JAX release date after `D`, when we release version `0.4.31`):
+    we start the clock for the 6 months backwards compatibility
```

Comment:
This implies that we need at least ~7 months to fully remove a custom call. I can see the logic here, but I can't help but feel that this is long.

Curious what Peter and Yash would say about this.

Reply:
@hawkinsp @yashk2810 

It is long, and I am afraid it will still make some users unhappy. I have heard from researchers that they want to re-run their experiments for more than 6 months. 

The real solution here is to rely less on custom calls, and push more logic into StableHLO and the compiler.